Automate API doc with CI

NicolasMassart opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/teku/blob/master/CONTRIBUTING.md -->

## PR Description
This is the second attempt to provide CI automated API doc published on Github Pages.
First attempt failed for many reasons and mostly because I faced issues with pushing changes on a protected branch. Now this one uses a gh-pages specific branch and no generated content is stored on the master. Only source files and scripts for doc publishing are on master (or in this branch for now). This was refined in other projects like we3js-eea or orchestrate node SDK before.

For now, on this PR, you will be able to notice the publication in a [test branch](https://github.com/PegaSysEng/teku/tree/test-gh-pages) but it will be similar in [the real one](https://github.com/PegaSysEng/teku/tree/gh-pages).

~~**IMPORTANT:** the PR is draft because it is required to remove Circle CI `GH_PAGES_BRANCH_OVERRIDE` env var and the `publishAPIDoc` job workflow filter for `1826-api-doc-on-CI` before merge. So draft should work like a warning and prevent merging without taking care of this.~~

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->
fixes #1826

NicolasMassart committed:
update scripts and readme

NicolasMassart committed:
remove version file that will be in gh-pages

NicolasMassart committed:
update CI

NicolasMassart committed:
testing CI filter

NicolasMassart committed:
fix artifact path

NicolasMassart committed:
fix cache and js script

NicolasMassart committed:
fix wrong dir

NicolasMassart committed:
Merge branch 'master' into 1826-api-doc-on-CI

ajsutton commented on .circleci/config.yml:
```diff
@@ -307,53 +307,88 @@ jobs:
-            git config --global user.name $CIRCLE_USERNAME
-            git config --global user.email "${CIRCLE_USERNAME}@users.noreply.github.com"
+            # publishing happens by pushing files to gh-pages branch
+            # retrieve branch to publish to from CIrcle CI env vars and fallback on test-gh-pages
```
nit:
```suggestion
            # retrieve branch to publish to from Circle CI env vars and fallback on test-gh-pages
```

ajsutton commented on .circleci/config.yml:
```diff
@@ -307,53 +307,88 @@ jobs:
+            # publishing happens by pushing files to gh-pages branch
+            # retrieve branch to publish to from CIrcle CI env vars and fallback on test-gh-pages
+            # if no var defined
+            [ -n "$BRANCH_OVERRIDE" ] && readonly BRANCH="$BRANCH_OVERRIDE" || readonly BRANCH='test-gh-pages'
```
I'd suggest we should fall back to `gh-pages` so that the code says what it will do in the normal case.  Removing the `BRANCH_OVERRIDE` is probably ideal if we can so that you don't have to go digging through CircleCI to find out what this will actually do.

NicolasMassart replied:
This code was suggested by @benjamincburns so I would like his opinion on that.

benjamincburns replied:
I agree with @ajsutton - his approach is better than the one I suggested.

ajsutton commented on .circleci/config.yml:
```diff
@@ -307,53 +307,88 @@ jobs:
+            # copy the content of the docs folder built on the previous job and attached
+            # using workspace on temp location
+            # cp with T option to override existing content, specially usefull for latest
+            cp -aT /tmp/workspace/docs/. .
```
Should we delete the existing content first?  Otherwise if a file is moved both versions will stay around forever with one being stale.

NicolasMassart replied:
We want to keep the content to preserve all the specs from previous versions. If any changes are made that require to delete some files from gh-pages branch, it would be done by someone knowing the system and who will be able to update the gh-pages branch manually. It should not be a process that will be used often, so I prefer not to build a complex thing and potential bugs for that.

ajsutton reviewed (commented):
LGTM.  Would be keen for a Yeti thumbs up as they're more familiar with this kind of thing than I am.

NicolasMassart committed:
typo

NicolasMassart committed:
updated branch name override per review comment

and added a more explicit name as otherwise we don't know what branch
override it is about.

NicolasMassart committed:
Merge branch 'master' into 1826-api-doc-on-CI

NicolasMassart committed:
remove test publishing
